dff/script/slots/types.py

```diff
@@ -0,0 +1,300 @@
+"""
+Types
+-----
+This module encapsulates different types of slots.
+"""
+import re
+from abc import ABC, abstractmethod
+from copy import copy
+from collections.abc import Iterable
+from typing import Callable, Any, Tuple, Dict, Union, Optional
+
+from pydantic import Field, BaseModel, field_validator
+
+from dff.script import Context
+from dff.pipeline.pipeline.pipeline import Pipeline, SLOT_STORAGE_KEY
+from .utils import singleton
+
+
+class BaseSlot(BaseModel, ABC, arbitrary_types_allowed=True):
+    """
+    BaseSlot is a base class for all slots.
+    Not meant for direct subclassing, unlike :py:class:`~.ValueSlot` and :py:class:`~._GroupSlot`.
+    """
+
+    name: str
+    children: Optional[Dict[str, "BaseSlot"]]
+
+    @field_validator("name", mode="before")
+    def validate_name(cls, name: str):
+        if "/" in name:
+            raise ValueError("Character `/` cannot be used in slot names.")
+        return name
+
+    def __init__(self, name: str, **data):
+        super().__init__(name=name, **data)
+
+    def __deepcopy__(self, *args, **kwargs):
+        return copy(self)
+
+    def __eq__(self, other: "BaseSlot"):
+        return self.dict(exclude={"name"}) == other.dict(exclude={"name"})
+
+    def has_children(self) -> bool:
+        return self.children is not None and len(self.children) > 0
+
+    @abstractmethod
+    def unset_value(self) -> Callable[[Context, Pipeline], None]:
+        raise NotImplementedError("Base class has no attribute 'value'")
+
+    @abstractmethod
+    def get_value(self) -> Callable[[Context, Pipeline], Dict[str, Union[str, None]]]:
+        raise NotImplementedError("Base class has no attribute 'value'")
+
+    @abstractmethod
+    def is_set(self) -> Callable[[Context, Pipeline], bool]:
+        raise NotImplementedError("Base class has no attribute 'value'")
+
+    @abstractmethod
+    def fill_template(self, template: str) -> Callable[[Context, Pipeline], str]:
+        raise NotImplementedError("Base class has no attribute 'value'")
+
+    @abstractmethod
+    def extract_value(self, ctx: Context, pipeline: Pipeline) -> Any:
+        """
+        `Extract value` method is distinct for most slots. So, if you would like to
+        introduce your own slot type, it is assumed, that you will override the
+        extract_value method.
+        """
+        raise NotImplementedError("Base class has no attribute 'value'")
+
+
+class _GroupSlot(BaseSlot):
+    """
+    Base class for :py:class:`~.RootSlot` and :py:class:`~.GroupSlot`.
+
+    """
+
+    value: None = Field(None)
+    children: Dict[str, BaseSlot] = Field(default_factory=dict)
+
+    @field_validator("children", mode="before")
+    def validate_children(cls, children, values: dict):
```

Comment:
`children` is assumed to be an instance of `Iterable`, it's not checked if it can be used with `len()` function.
```suggestion
    def validate_children(cls, children: Iterable, values: Dict) -> Dict[str, BaseSlot]:
```